### DISCUSSION TITLE
`null` value in nested field doesn't get updated

### DISCUSSION THREAD

**Discussion by okbytes** [react-hook-form/react-hook-form#5752] [Q&A] [answered]
**Describe the bug**
I have a nested piece of form configuration that I cannot figure out how to update. It comes down from the API as `null` if it's never been set. I can update the `__typename` of `pagination` just fine and see the value populate in the `Listbox` component. However, the parent `pagination` configuration remains `null`. The dynamic form pieces that key on `__typename` remain hidden and unable to be configured.

I tried setting the `configuration.pagination` parent on my `Listbox` instead of `pagination.__typename` but got more strange behavior. The form pieces would render, but all the text inputs would fail to show the values when typing. `console.log` only revealed those values as a single character not as all of them as I would expect from a text input. Using `setValue` didn't work either as it would give me TypeScript errors about being unable to assign to type `never`.

**To Reproduce**
- Choose a type from the empty select (there are only two)
- The rest of the pagination still shows `null`, and so the dynamic form pieces don't render

**Codesandbox link (Required)**
https://codesandbox.io/s/infallible-galois-vdbou?file=/src/App.tsx

V7.7.1

**Expected behavior**
When setting the `pagination.__typename` of a `pagination: null`, I would expect `pagination` to be `pagination: { __typename: value }`.

**Desktop (please complete the following information):**

- OS: macOS 11.4 on Apple Silicon
- Browser: Chrome
- Version: 91.0.4472.114 (Official Build) (arm64)

**Comment by bluebill1049** [comment]
This is not clear to me how to reproduce the issue.
<img width="335" alt="Screen Shot 2021-06-28 at 12 28 01 pm" src="https://user-images.githubusercontent.com/10513364/123570779-485a0b80-d80c-11eb-8fc5-579ac24b3491.png">

> Choose a type from the empty select (there are only two)
> The rest of the pagination still shows null, and so the dynamic form pieces don't render

**Comment by bluebill1049** [comment]
and please stripe above that additional logic and component to help me to investigate the problem.

**Comment by okbytes** [comment]
I've simplified the CSB here: https://codesandbox.io/s/intelligent-pascal-ojpz4?file=/src/App.tsx

Let me know what doesn't make sense and I can try to strip out more.

The main problem is I have a property that will render two different pieces of form depending on what's selected. In this example, `pagination.__typename` will determine whether a `text` input is rendered underneath or a `textarea`. The problem I'm having is when the `select` makes a choice, it is not updated in the form. I've put `console.log` to show that the watched `paginationType` remains `null`.

**Comment by okbytes** [comment]
Sorry, Bill. I had a mistake in the CSB. I've saved it now. It shows my problem but differently. I can change the parent component to render the specific children, but the children don't hold their values like a normal text/textarea.

**Comment by bluebill1049** [comment]
thanks @01brett can you share the steps to reproduce the issue?

**Comment by okbytes** [comment]
Sure thing.

- Click the "type" button and select one of the two options
- Confirm the button now has text in it
- A text field or textarea should appear beneath it (depending on the choice)
- Try to type in the text field/textarea
- Confirm you cannot see any characters you've typed
- Note: the console will log when you type. Part of the problem is it logs one character at a time and you are unable to see any characters in the text input/textarea

**Comment by bluebill1049** [comment]
<img width="1028" alt="Screen Shot 2021-06-28 at 1 23 59 pm" src="https://user-images.githubusercontent.com/10513364/123574942-19e02e80-d814-11eb-986d-2c64ae5d4833.png">

**Comment by okbytes** [comment]
Ah, shoot. Sorry! I added a `?` to the lookup. If you refresh, it should work now.

**Comment by bluebill1049** [comment, marked as answer]
please move it outside of the Controller

**Comment by bluebill1049** [comment]
```tsx
function NestedFormPiece(props: Props) {
  const { control, register } = props.methods;

  return (
    <>
      <Controller
        control={control}
        name="configuration.pagination"
        render={({ field: parentField }) => {
          return (
            <>
              <label style={{ display: "block" }}>url</label>
              <input
                style={{ display: "block" }}
                {...register("configuration.url")}
              />
              <Controller
                control={control}
                name="configuration.pagination.__typename"
                render={({ field }) => {
                  return (
                    <div className="relative">
                      <Listbox
                        value={field.value}
                        onChange={(value) => {
                          parentField.onChange({ __typename: value });
                          field.onChange(value);
                        }}
                      >
                        <Listbox.Label style={labelStyle}>Type</Listbox.Label>
                        <Listbox.Button
                          style={{ height: "32px", minWidth: "100px" }}
                        >
                          {field.value}
                        </Listbox.Button>
                        <Listbox.Options style={optsStyles}>
                          <Listbox.Option
                            style={optStyle}
                            value="NextPagePagination"
                          >
                            Next Page name
                          </Listbox.Option>
                          <Listbox.Option
                            style={optStyle}
                            value="TokenPagination"
                          >
                            Token
                          </Listbox.Option>
                        </Listbox.Options>
                      </Listbox>
                    </div>
                  );
                }}
              />
            </>
          );
        }}
      />

      <div>
        <label style={{ display: "block" }}>nextpage name</label>
        <textarea
          style={{ display: "block" }}
          {...register("configuration.pagination.nextpagename")}
        />
      </div>
    </>
  );
}
```

prettier much mixed controlled with uncontrolled.
